### ISSUE TITLE
Nonlinear operator norm

### ISSUE THREAD

**Issue by adler-j** [odlgroup/odl#551]
It is often of interest to find the nonlinear operator norm, and this is especially relevant for things like:

```
odl.diagnostics.OperatorTest(op).derivative()
```

which currently throws an error if `op` is nonlinear:

```
odl.operator.operator.OpNotImplementedError: adjoint not implemented for operator ...
```

I suggest we add a nonlinear utility, similar to `power_method_opnorm` that computes the operator norm for nonlinear operators. In this case, I'd define the operator norm as the Lipschitz constant of the operator.

```
min a s.t. ||A(x) - A(y)|| <= a||x - y||
```

In addition to this, we could add a new front facing method called `opnorm` which calls `power_method_opnorm` if the operator is linear, and otherwise `opnorm_lipschitz`. We could also do a rename `power_method_opnorm -> opnorm_power_method` to help autocomplete.

**Comment by kohr-h** [comment]
Sounds nice. Would you compute it with brute force or do you know a better algorithm? Agree to the renaming.

**Comment by kohr-h** [comment]
No, and google-scholaring doesn't return any useful results apart from the one you linked to. I would guess it's a more sophisticated version of our brute force approach, but nothing really intelligent.

Bouncing the question to @ozanoktem also.

**Comment by adler-j** [comment]
Idea for operators that implement `derivative` is to simply use it:

```python
def opnorm_lipschitz(operator):
    return max(opnorm_power_method(op.derivative(ex)) for ex in op.domain.examples)
```

Another option is to interpret the problem as that of finding a minimizer to

```
min_[x, y] - ||A(x) - A(y)|| / ||x - y||
```

which should be possible to solve using a nonlinear minimization algorithm.

**Comment by kohr-h** [comment]
That's of course possible. Actually, the `derivative` test doesn't make use of the operator norm, so there is no "circular reasoning" danger here.

**Comment by kohr-h** [comment]
> Another option is to interpret the problem as that of finding a minimizer to

maximizer

But I can imagine that this problem is strongly underdetermined, so there are probably lots of local maxima. I'm more convinced of some "maximum over random sample" or a slightly more intelligent approach of the same kind.

**Comment by adler-j** [comment]
> maximizer

Actually minimizer since i snuck in a `-` there :smile:.

> But I can imagine that this problem is strongly underdetermined, so there are probably lots of local maxima. I'm more convinced of some "maximum over random sample" or a slightly more intelligent approach of the same kind.

Probably correct.

**Comment by kohr-h** [comment]
> Actually minimizer since i snuck in a - there :smile:.

Ah don't do that :angry:
